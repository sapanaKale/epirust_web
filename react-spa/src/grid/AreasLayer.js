/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import React, { useContext, useRef, useState, useEffect } from 'react';
import { GridContext } from './index'
import { AreaColors } from './constants';

export default function AreasLayer({ areaDimensions }) {

    const { cellDimension, lineWidth, canvasDimension, size } = useContext(GridContext);

    const areasLayerCanvas = useRef(null);
    const [areasCanvasContext, setAreasCanvasContext] = useState(null);

    useEffect(() => {
        if (!areaDimensions)
            return

        if (!areaDimensions || !areasLayerCanvas)
            return

        setAreasCanvasContext(areasLayerCanvas.current.getContext("2d"));

    }, [areasLayerCanvas, areaDimensions])

    useEffect(() => {
        if (!areaDimensions || !areasCanvasContext)
            return

        function updateAreaColor(areaDimensions, x, y) {
            let area;
            for (var i = 0; i < areaDimensions.length; i++) {
                if (isWithinArea(areaDimensions[i], x, y)) {
                    area = areaDimensions[i];
                    break;
                }
            }

            const color = area ? area.color : AreaColors.OTHER;

            if (areasCanvasContext.fillStyle !== color) {
                areasCanvasContext.fillStyle = color;
            }

            // areasCanvasContext.fillStyle = "white";
        }

        for (let x = 0; x < size; x++) {
            for (let y = 0; y < size; y++) {
                updateAreaColor(areaDimensions, x, y);
                areasCanvasContext.fillRect((x * cellDimension) + lineWidth / 2, (y * cellDimension) + lineWidth / 2, cellDimension, cellDimension)
            }
        }
    }, [areasCanvasContext, size, cellDimension, lineWidth, areaDimensions])

    return (
        <canvas ref={areasLayerCanvas} data-testid="grid-canvas" id="grid-canvas" width={canvasDimension} height={canvasDimension} style={{ position: "absolute", zIndex: 1 }} />
    )
}

function isWithinArea(area, currentX, currentY) {

    return (currentX >= area.start_offset.x && currentY >= area.start_offset.y
        && currentX <= area.end_offset.x && currentY <= area.end_offset.y);
}
