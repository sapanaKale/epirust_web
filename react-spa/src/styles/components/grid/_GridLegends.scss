.legends-wrap {
    border: 2px solid powderblue;
    position: absolute;
    top: 240px;
    right: 32px;

    .legends {
        padding: 5px;
        list-style: none;

        .legend-item {
            margin-bottom: 5px;
        }

        .swatch {
            margin-right: 4px;
            display: inline-block;
        }
        .text {
            vertical-align: text-bottom;
            font-size: 12px;
        }

        .count {
            padding-right: 4px;
        }


        &.areas {
            .swatch {
                padding: 8px 15px;
            }
        }

        &.agents {
            .swatch {
                padding: 5px;
                border-radius: 50%;
            }
        }
    }
}
