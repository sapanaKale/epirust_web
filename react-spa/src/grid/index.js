/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from 'react';

import LandmarksLayer from './LandmarksLayer';
import LinesLayer from './LinesLayer';
import AreasLayer from './AreasLayer';
import AgentsLayer from './AgentsLayer';
import {AreaColors} from './constants';
import io from 'socket.io-client'
import config from "../config";
import Loader from '../common/Loader';
import PropTypes from "prop-types";

export const GridContext = React.createContext(null);

export default function GridPage({jobId}) {
    const [socket, setSocket] = useState(null);
    const [socketDataExhausted, setSocketDataExhausted] = useState(false);

    //default values?
    const [areaDimensions, setAreaDimensions] = useState(null);
    const [landmarksDimensions, setLandmarksDimensions] = useState(null);
    const [agentPositions, setAgentPositions] = useState(null);

    const [gridContextData, setGridContextData] = useState(null);

    useEffect(() => {
        console.log("started socket");
        const socketInstance = io(`${config.API_HOST}/grid-updates`);
        setSocket(socketInstance);

        return () => {
            socketInstance.close();
        }
    }, []);

    //reading socket data
    useEffect(() => {
        if (!socket)
            return;

        socket.emit('get', {jobId});

        socket.on('gridData', function (message) {
            if ("simulation_ended" in message) {
                socket.close();
                setSocketDataExhausted(true);
                return
            }

            if ('grid_size' in message) {
                const { housing_area, work_area, transport_area, hospital_area, grid_size } = message;
                const areaDimensions = [
                    { ...housing_area, color: AreaColors.HOUSING },
                    { ...work_area, color: AreaColors.WORK },
                    { ...transport_area, color: AreaColors.TRANSPORT },
                    { ...hospital_area, color: AreaColors.HOSPITAL }
                ];

                const housesDimensions = message.houses,
                    officesDimensions = message.offices;

                const cellDimension = Math.floor((window.innerHeight - 150) / grid_size),
                    lineWidth = Math.floor(cellDimension / 4) < 1 ? 0 : Math.floor(cellDimension / 4),
                    canvasDimension = (grid_size * cellDimension) + lineWidth;

                setGridContextData({
                    cellDimension: cellDimension,
                    lineWidth: lineWidth,
                    canvasDimension: canvasDimension,
                    size: grid_size
                });

                setAreaDimensions(areaDimensions);
                setLandmarksDimensions({ housesDimensions, officesDimensions });
                return
            }

            if ('citizen_states' in message) {

                setAgentPositions(pos => {
                    if (!pos)
                        return [message.citizen_states];

                    return [...pos, message.citizen_states]
                })
            }

        });
    }, [socket, jobId]);

    if (!gridContextData)
        return (
          <div className="grid-wrap">
              <Loader />
          </div>
        )

    return (
        <div className="grid-wrap">
            <GridContext.Provider value={gridContextData}>
                <div style={{ position: "relative" }}>
                    <AreasLayer areaDimensions={areaDimensions} />
                    <LinesLayer />
                    <LandmarksLayer landmarksDimensions={landmarksDimensions} />
                    {<AgentsLayer agentPositions={agentPositions} simulationEnded={socketDataExhausted} />}
                </div >
            </GridContext.Provider>
        </div>
    )
}

GridPage.propTypes = {
    jobId: PropTypes.string.isRequired
}