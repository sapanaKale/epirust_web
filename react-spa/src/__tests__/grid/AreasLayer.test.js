/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from 'react'
import { render } from '@testing-library/react'
import AreasLayer from '../../grid/AreasLayer'
import { GridContext } from '../../grid/index'

test('should render AreasLayer component', () => {
    const cellDimension = 2
    const lineWidth = 1
    const canvasDimension = 20
    const size = 10

    const areaDimensions = {
        "housing_area": {
            "start_offset": {
                "x": 0,
                "y": 0
            },
            "end_offset": {
                "x": 4,
                "y": 10
            }
        },
        "work_area": {
            "start_offset": {
                "x": 5,
                "y": 0
            },
            "end_offset": {
                "x": 7,
                "y": 10
            }
        },
        "transport_area": {
            "start_offset": {
                "x": 4,
                "y": 0
            },
            "end_offset": {
                "x": 5,
                "y": 10
            }
        },
        "hospital_area": {
            "start_offset": {
                "x": 7,
                "y": 0
            },
            "end_offset": {
                "x": 8,
                "y": 10
            }
        },
    }

    const { getByTestId } = render(<GridContext.Provider value={{ cellDimension, lineWidth, canvasDimension, size }}><AreasLayer areaDimensions={areaDimensions} /></GridContext.Provider>)
    const context = getByTestId("grid-canvas").getContext("2d")
    expect(context.__getEvents()).toMatchSnapshot()
})