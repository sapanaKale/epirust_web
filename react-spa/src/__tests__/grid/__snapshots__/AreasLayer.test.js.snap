// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AreasLayer component 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#ccc",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 0.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 2.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 4.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 6.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 8.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 10.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 12.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 14.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 16.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 0.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 2.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 4.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 6.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 8.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 10.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 12.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 14.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 16.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 2,
      "width": 2,
      "x": 18.5,
      "y": 18.5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;
